[MLv2] Round-trip JS <-> CLJS Cards on demand; use for display locking

bshepherdson opened the pull request:
This introduces a Malli schema for `Card`, the representation for a
"question" in Metabase, and supports conversion between the JS map
`Card` and CLJS maps.

The `dataset_query` field holding the native or MBQL query is special,
and this PR uses the `->legacy-MBQL` conversion from #29419.

A TS `Question` can be created with either the JS or CLJS rep for the
card, and the methods `Question.card()` and `Question._cljsCard()`
lazily convert the other form into the requested form and cache it.

The existing `_getMLv2Query` logic is replaced; it is subsumed by the
conversion of the whole card including the query.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29520)
<!-- Reviewable:end -->

bshepherdson committed:
[MLv2] Round-trip JS <-> CLJS Cards on demand; use for display locking

This introduces a Malli schema for `Card`, the representation for a
"question" in Metabase, and supports conversion between the JS map
`Card` and CLJS maps.

The `dataset_query` field holding the native or MBQL query is special,
and this PR uses the `->legacy-MBQL` conversion from #29419.

A TS `Question` can be created with either the JS or CLJS rep for the
card, and the methods `Question.card()` and `Question._cljsCard()`
lazily convert the other form into the requested form and cache it.

The existing `_getMLv2Query` logic is replaced; it is subsumed by the
conversion of the whole card including the query.

kulyk commented on frontend/src/metabase-lib/Question.ts:
```diff
@@ -99,16 +101,67 @@ export type QuestionCreatorOpts = {
+  }
+
+  const ret = {};
+  for (const k of Object.keys(x)) {
+    ret[k] = massageForRoundTripCheck(x[k]);
+  }
```
Absolute nit: we usually prefer `forEach` instead of a `for`

kulyk commented on frontend/src/metabase-lib/Question.ts:
```diff
@@ -99,16 +101,67 @@ export type QuestionCreatorOpts = {
+  return ret;
+}
+
+function diffKeys(a, b) {
+  const keys = [];
+  for (const k of Object.keys(a)) {
+    if (!_.isEqual(a[k], b[k])) {
+      keys.push(k);
+    }
+  }
+  return keys;
+}
```
Think we can use [`_.difference`](https://underscorejs.org/#difference) instead?

kulyk commented on frontend/src/metabase-lib/Question.ts:
```diff
@@ -152,18 +214,76 @@ class QuestionInner {
+          : undefined),
+      });
+      if (!_.isEqual(originalCard, roundTripped)) {
+        console.log(
+          originalCard,
+          roundTripped,
+          diffKeys(originalCard, roundTripped),
+        );
```
Could you remove the console.log please?

kulyk reviewed (commented):
The FE part looks good! Left some minorish comments

camsaul commented on src/metabase/domain_entities/card.cljc:
```diff
@@ -0,0 +1,260 @@
+    [:creator_id {:optional true} number?]
+    [:creation-type {:optional true :js/prop "creationType"} string?]
+    [:dashboard_count {:optional true} number?]
+    [:dashboard-id {:optional true :js/prop "dashboardId"} number?]
```
It seems a little bit weird to mix and match kebab-case and snake_case, if we're already converting JS names from JS to kebeb-case, can we also convert names in Clojure-land? If not I'm worried that this would be hard to use in JVM Clojure, because we're still using `:dashboard_id` there

bshepherdson replied:
The mixing of names here is really unfortunate, I agree. There's issues with schemes to try to clean this up, though.

- This isn't actually the `dashboard_id` field from the appdb. It's attached dynamically in the client sometimes to keep track of the parent dashboard. (Hence the JS-convention name `dashboardId`.)
    - In practice it's not even used, it's just there so the conversion of `Card`s is done correctly in JS-land.
- There's not currently any point where conversion is happening in Clojure-land.

That said, for fields like this that we don't actually expect to see on the BE side right now, we could spell this specific one as `:database_id` just for minimal surprise.

camsaul commented on src/metabase/domain_entities/card.cljc:
```diff
@@ -0,0 +1,260 @@
+    [:visualization_settings VisualizationSettings]]])
+
+#?(:cljs
+   (def ^:private CardNoQuery
+     "Modified schema for Card, with the `:dataset_query` key set to :any.
+     Queries are special and we want Malli to disregard them."
+     (-> Card
+         mc/schema
+         mc/children
+         first
+         (mut/assoc-in [:dataset_query] :any))))
```
Could you have `CardNoQuery` be the base version and then have `Card` just be that plus `mu/merge` the schema that includes `:dataset_query`? Seems like it would mean you'd need to dig into Malli internals a little less

camsaul reviewed (approved):
Backend stuff works for me, I'll defer to @kulyk RE frontend changes. Is there a way we can test any of the domain entity convertor things in Cljc? If we break one it would be nice to possibly get feedback before waiting for FE unit or e2e tests to fail

bshepherdson committed:
review commentary

bshepherdson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520)* status: 0 of 18 files reviewed, 8 unresolved discussions (waiting on @camsaul and @kulyk)

___
*[`frontend/src/metabase-lib/Question.ts` line 116 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-116:-NRcE_e56yM9TPbrK9SI:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L116)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

`undefined == null`, so we should be able to avoid the `undefined` check on the left
</blockquote></details>

Done.
___
*[`frontend/src/metabase-lib/Question.ts` line 119 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-119:-NRcEa6kA8n9TRnoq06U:bfw546s) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L119)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

Would it make sense to type the `x` argument to make sure it won't receive an object?
</blockquote></details>

No, this is intended to run recursively on all the values of objects and all the elements of arrays, so sometimes `x` is a string or number and should just be returned as-is.
___
*[`frontend/src/metabase-lib/Question.ts` line 127 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-127:-NRcF6jj2zu461j07XAA:b813s9m) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L127)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

What does `ret` mean here? I'm also not sure what exactly is going on here, and which case is handled here. Would it make sense to add a comment maybe? On the other hand, maybe giving this condition a name would help?
</blockquote></details>

I've added comments. `ret` is a personal convention for "new structure I'm creating so I can return it shortly", but it's not very transparent. It's called `result` now.
___
*[`frontend/src/metabase-lib/Question.ts` line 132 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-132:-NRcG8Sw5mGgEbJcK72Y:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L132)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

Absolute nit: we usually prefer `forEach` instead of a `for`
</blockquote></details>

Done.
___
*[`frontend/src/metabase-lib/Question.ts` line 144 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-144:-NRcFf1Z-jwK1PCl4tN1:bfmbeos) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L144)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

Think we can use [`_.difference`](https://underscorejs.org/#difference) instead?
</blockquote></details>

Unfortunately not. `_.difference` works on arrays, no objects. We could use it to check if the key sets are the same, but we also want to check that the values at those keys are identical.

Anyway, `diffKeys` is really a debugging helper we don't need to preserve. I've removed it.
___
*[`frontend/src/metabase-lib/Question.ts` line 251 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TPA6Thkzman4Nq-r1-251:-NRcGFW15J2TGdmt99Jl:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L251)):*
<details><summary><i>Previously, kulyk (Anton Kulyk) wrote…</i></summary><blockquote>

Could you remove the console.log please?
</blockquote></details>

Done.
___
*[`src/metabase/domain_entities/card.cljc` line 134 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TQhV576COvCfJN-r1-134:-NRcHvz9ElG4YWlbkwBR:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/src/metabase/domain_entities/card.cljc#L134)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

Could you have `CardNoQuery` be the base version and then have `Card` just be that plus `mu/merge` the schema that includes `:dataset_query`? Seems like it would mean you'd need to dig into Malli internals a little less
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

camsaul reviewed (changes requested):
Left a few notes

> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520#-:-NRjEK3u6B4XhGAYnrVP:b-edp8hm)* status: all files reviewed, 11 unresolved discussions (waiting on @bshepherdson and @kulyk)

___
*[`package.json` line 295 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjCKoG9fcxdkV8BTj6:-NRjCKoH6F4WtVOvPQYg:b2jfk7j) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/package.json#L295)):*
> ```JSON
>     "test-unit-keep-cljs": "jest --maxWorkers=2 --config jest.unit.conf.json",
>     "test-unit-watch": "yarn test-unit --watch",
>     "test-unit-watch-keep-cljs": "jest --maxWorkers=2 --config jest.unit.conf.json --watch",
> ```

You can just do 

```sh
yarn test-unit-keep-cljs --watch
```

That's what I've been doing. Do we really need a separate command for every possible combination of options?
___
*[`frontend/src/metabase-lib/Question.ts` line 25 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjCbt165fT6F-lMVvZ:-NRjCbt165fT6F-lMVv_:b-491pnk) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L25)):*
> ```XML
> import { isFK } from "metabase-lib/types/utils/isa";
> import { memoizeClass, sortObject } from "metabase-lib/utils";
> import * as C from "cljs/metabase.domain_entities.card";
> ```

There's no way reading `C.whatever` could ever be confusing, right? :laughing: 
___
*[`frontend/src/metabase-lib/metadata/Table.ts` line 61 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjD4Nx2eBvtYis1_EH:-NRjD4Nx2eBvtYis1_EI:undefined) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/metadata/Table.ts#L61)):*
> ```XML
>   question() {
>     return Question.create({
>       databaseId: (this.db && this.db.id) || this.db_id,
> ```





_Code quote (i):_
```XML
(this.db && this.db.id)
```



_Code snippet (ii):_
```XML
this.db?.id
```
___
*[`resources/sample-database.db.mv.db` line 0 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjEG2B5eyiDQz2_ReH:-NRjEG2B5eyiDQz2_ReI:bl0wu5w) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/resources/sample-database.db.mv.db#L0)):*
Accidentally committed the binary file again
___
*[`src/metabase/domain_entities/card.cljc` line 92 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRcE6TQhV576COvCfJN-r1-92:-NRjDJBV8R0s0N0r2oB5:b32wtd) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/src/metabase/domain_entities/card.cljc#L92)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

The mixing of names here is really unfortunate, I agree. There's issues with schemes to try to clean this up, though.

- This isn't actually the `dashboard_id` field from the appdb. It's attached dynamically in the client sometimes to keep track of the parent dashboard. (Hence the JS-convention name `dashboardId`.)
    - In practice it's not even used, it's just there so the conversion of `Card`s is done correctly in JS-land.
- There's not currently any point where conversion is happening in Clojure-land.

That said, for fields like this that we don't actually expect to see on the BE side right now, we could spell this specific one as `:database_id` just for minimal surprise.
</blockquote></details>

ok
___
*[`src/metabase/lib/js.cljs` line 50 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjDwJP3Q7ej4b9sXLv:-NRjDwJP3Q7ej4b9sXLw:blc2iau) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/src/metabase/lib/js.cljs#L50)):*
> ```Clojure
>   "Coerce a CLJS pMBQL query back to (1) a legacy query (2) in vanilla JS."
>   [query-map]
>   (-> query-map convert/->legacy-MBQL clj->js))
> ```

You almost certainly want to be doing `(clj->js :keyword-fn u/qualified-name)` here.



_Code quote (i):_
```Clojure
clj->js
```



_Code snippet (ii):_
```Clojure
(clj->js :keyword-fn u/qualified-name)
```


<!-- Sent from Reviewable.io -->

bshepherdson committed:
Review comments, text fixes

bshepherdson committed:
integration with earlier changes. add v2 native.ts and use it

bshepherdson committed:
quick fix

bshepherdson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520)* status: 1 of 25 files reviewed, 10 unresolved discussions (waiting on @camsaul and @kulyk)

___
*[`frontend/src/metabase-lib/Question.ts` line 25 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjCbt165fT6F-lMVvZ:-NRkBoHdAHwOUyESRrSS:b-zif2or) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/frontend/src/metabase-lib/Question.ts#L25)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

There's no way reading `C.whatever` could ever be confusing, right? :laughing: 
</blockquote></details>

I've replaced it with `CARD`, following the lead of `as AGGREGATION` etc. in this file. I'm open to whatever prefix. But this is also dead code walking.
___
*[`src/metabase/lib/js.cljs` line 50 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjDwJP3Q7ej4b9sXLv:-NRkCmJ61b0ozJrxtC7s:b-4b3zbn) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/src/metabase/lib/js.cljs#L50)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

You almost certainly want to be doing `(clj->js :keyword-fn u/qualified-name)` here.
</blockquote></details>

Okay. (That doesn't remove the need for `fix-namespaced-values`, though, since `:keyword-fn` only applies to map keys and not values.)
___
*[`package.json` line 295 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjCKoG9fcxdkV8BTj6:-NRkB58x1xGJNXjsQshJ:bk9kyqz) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/package.json#L295)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

You can just do 

```sh
yarn test-unit-keep-cljs --watch
```

That's what I've been doing. Do we really need a separate command for every possible combination of options?
</blockquote></details>

Fair enough, I've dropped it.
___
*[`resources/sample-database.db.mv.db` line  at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjEG2B5eyiDQz2_ReH:-NRkCHhD5wGHTLAT3awn:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/resources/sample-database.db.mv.db#L)):*
<details><summary><i>Previously, camsaul (Cam Saul) wrote…</i></summary><blockquote>

Accidentally committed the binary file again
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bshepherdson committed:
lint

camsaul reviewed (approved):
> Reviewed 24 of 24 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520)* status: all files reviewed, 9 unresolved discussions (waiting on @bshepherdson and @kulyk)

___
*[`frontend/src/metabase-lib/metadata.ts` line 11 at r2](https://reviewable.io/reviews/metabase/metabase/29520#-NRtIudx4z6d8Oiyuh5O:-NRtIudx4z6d8Oiyuh5P:b-3kv81s) ([raw file](https://github.com/metabase/metabase/blob/21441982164078678d8740b639ecf6b450df46ad/frontend/src/metabase-lib/metadata.ts#L11)):*
> ```XML
>   metadata: Metadata,
> ): MetadataProvider {
>   return ML.__GT_metadata_provider(databaseId, metadata);
> ```

This is foul, maybe we should just rename the underlying function `to-metadata-provider` or `as-metadata-provider`, idk.
___
*[`frontend/test/__support__/sample_database_fixture.json` line 65 at r2](https://reviewable.io/reviews/metabase/metabase/29520#-NRtJBnF6Yt3h_uWbGBE:-NRtJBnF6Yt3h_uWbGBF:b-e6vm7x) ([raw file](https://github.com/metabase/metabase/blob/21441982164078678d8740b639ecf6b450df46ad/frontend/test/__support__/sample_database_fixture.json#L65)):*
> ```JSON
>         "table_id": 1,
>         "definition": {
>           "filter": [">", ["field", 6, null], 20],
> ```

FWIW technically this old syntax was allowed in legacy-legacy-legacy MBQL, the `:field` clause is supposed to be implied as the first arg in filters in really old MBQL. If you run this thru `mbql.normalize/normalize` it will fix it. So if you needed to change this, this is probably a sign that something wasn't ran thru normalize that should have been before doing `->pMBQL` or whatever
___
*[`frontend/test/metabase-lib/lib/Question.unit.spec.js` line 73 at r2](https://reviewable.io/reviews/metabase/metabase/29520#-NRtJcgV2_DLw0DxJppD:-NRtJcgV2_DLw0DxJppE:b-qn2j6c) ([raw file](https://github.com/metabase/metabase/blob/21441982164078678d8740b639ecf6b450df46ad/frontend/test/metabase-lib/lib/Question.unit.spec.js#L73)):*
> ```JavaScript
>       "source-table": ORDERS.id,
>       aggregation: [
>         ["count-where", [">", ["field", ORDERS.TOTAL.id, null], 50]],
> ```

Same as what I mentioned above
___
*[`src/metabase/lib/js.cljs` line 50 at r1](https://reviewable.io/reviews/metabase/metabase/29520#-NRjDwJP3Q7ej4b9sXLv:-NRtJuNl-We3q0A7i5P0:bmznp15) ([raw file](https://github.com/metabase/metabase/blob/843c7bfadaac17015228c78f02d9039095354be4/src/metabase/lib/js.cljs#L50)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

Okay. (That doesn't remove the need for `fix-namespaced-values`, though, since `:keyword-fn` only applies to map keys and not values.)
</blockquote></details>

It works on values too.

```clj
cljs.user> (clj->js {:my/namespace :my/namespace})
{"namespace" "namespace"}

cljs.user> (clj->js {:my/namespace :my/namespace} :keyword-fn metabase.util/qualified-name)
{"my/namespace" "my/namespace"}
```


<!-- Sent from Reviewable.io -->

bshepherdson committed:
Exclude dataset_query from the round trip comparison

This isn't ideal - it's often the problem - but it's also often
different in ways that don't actually matter but fail the check.
(Eg. dropping null/undefined keys)

bshepherdson committed:
Debugging changes - clean these up

camsaul reviewed (approved):
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520)* status: all files reviewed, 9 unresolved discussions (waiting on @bshepherdson and @kulyk)



<!-- Sent from Reviewable.io -->

metamben reviewed (approved):
I don't want to block, so <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520#-:-NS_4zzz1-M-ZamRiT5C:b1b3akw)* status: all files reviewed, 15 unresolved discussions (waiting on @bshepherdson and @kulyk)

___
*[`frontend/src/metabase-lib/Question.ts` line 383 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NSZztiL4HD_WU9x_fs9:-NSZztiL4HD_WU9x_fsA:b-2qdod) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L383)):*
> ```XML
>    * The visualization type of the question
>    */
>   display(): string {
> ```

Why isn't `undefined` listed here when it's listed for `cacheTTL()`? (I suppose the original should have listed `null` and we inherited the inconsistency.)
___
*[`frontend/src/metabase-lib/Question.ts` line 391 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NS_0lAy8NVKeqkd9nXm:-NS_0lAy8NVKeqkd9nXn:b-9n50g4) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L391)):*
> ```XML
>   }
> 
>   cacheTTL(): number | null {
> ```

`displayName()` list both `null` and `undefined`, here we list only `null`. Can we make them consistent?
___
*[`frontend/src/metabase-lib/Question.ts` line 411 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NS_-6be2RWEuDIbaZ3k:-NS_-6be2RWEuDIbaZ3l:b-1rylum) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L411)):*
> ```XML
>   }
> 
>   isPersisted() {
> ```

Why don't we list the type here? Just a historical accident?
___
*[`frontend/src/metabase-lib/Question.ts` line 437 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NS_2Ws98bbOg9go18Tw:-NS_2Ws98bbOg9go18Tx:bpbw0bk) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L437)):*
> ```XML
> 
>   displayIsLocked(): boolean {
>     return CARD.display_is_locked(this._cljsCard());
> ```

In most cases we get the value in JS, here in CLJS. How do we decide which way to go?
___
*[`frontend/src/metabase-lib/Question.ts` line 606 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NS_3Agl4RdVElpzuMeX:-NS_3Agl4RdVElpzuMeY:b-sav8nn) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L606)):*
> ```XML
> 
>   canWrite(): boolean {
>     return this.card()?.can_write;
> ```

Shouldn't this do the `!!` trick?
___
*[`frontend/src/metabase-lib/Question.ts` line 1063 at r3](https://reviewable.io/reviews/metabase/metabase/29520#-NS_3nuF1AKuxgGPdIRH:-NS_3nuF1AKuxgGPdIRI:bko7sir) ([raw file](https://github.com/metabase/metabase/blob/b76932cb6ab363f2054591644eb78a49d4a5f8cb/frontend/src/metabase-lib/Question.ts#L1063)):*
> ```XML
> 
>   description(): string | null {
>     return this.card().description;
> ```

Isn't a `?` missing here?


<!-- Sent from Reviewable.io -->

camsaul reviewed (approved):
Let's get this merged <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29520#-:-NSlaznm1I12f7QsRg_n:begfwy0)* status: all files reviewed, 16 unresolved discussions (waiting on @bshepherdson and @kulyk)

___
*[`test/metabase/lib/convert_test.cljc` line 205 at r4](https://reviewable.io/reviews/metabase/metabase/29520#-NSlaqfV3a_Rcq-vOJ-D:-NSlaqfV3a_Rcq-vOJ-E:b-fdht63) ([raw file](https://github.com/metabase/metabase/blob/9181381b6d4072939312fe295229ea3a4a8af473/test/metabase/lib/convert_test.cljc#L205)):*
> ```Clojure
>      :display "table"}))
> 
> (comment
> ```

Dumb question but is this `comment` because it doesn't work yet, or can it be added to the tests now if it does?


<!-- Sent from Reviewable.io -->
